apiserver/apiserver/web/challenge.py

```diff
@@ -0,0 +1,156 @@
+"""
+User challenge API endpoints - list user's challenges & issue new ones
+"""
+import datetime
+
+import flask
+import sqlalchemy
+
+from .. import model, util
+
+from . import match as match_api
+from . import util as api_util
+from .blueprint import web_api
+
+
+def make_challenge_record(challenge, participants):
+    result = {
+        "challenge_id": challenge["id"],
+        "time_created": challenge["created"],
+        "time_finished": challenge["finished"],
+        "num_games": challenge["num_games"],
+        "issuer": challenge["issuer"],
+        "winner": challenge["winner"],
+        "finished": bool(challenge["finished"]),
+        "players": {},
+    }
+
+    for participant in participants:
+        result["players"][participant["user_id"]] = {
+            "username": participant["username"],
+            "points": participant["points"],
+            "ships_produced": participant["ships_produced"],
+            "attacks_made": participant["attacks_made"],
+            "is_issuer": participant["user_id"] == result["issuer"],
+        }
+
+    return result
+
+
+def get_challenge_helper(challenge_id):
+    with model.engine.connect() as conn:
+        query = sqlalchemy.sql.select([
+            model.challenges.c.id,
+            model.challenges.c.created,
+            model.challenges.c.finished,
+            model.challenges.c.num_games,
+            model.challenges.c.issuer,
+            model.challenges.c.winner,
+        ]).select_from(model.challenges).where(
+            model.challenges.c.id == challenge_id
+        ).reduce_columns()
+
+        challenge = conn.execute(query).first()
+        if not challenge:
+            raise util.APIError(
+                404,
+                message="Challenge {} not found.".format(challenge_id))
+
+        participants = conn.execute(
+            model.challenge_participants.join(
+                model.users,
+                model.challenge_participants.c.user_id == model.users.c.id
+            ).select(
+                model.challenge_participants.c.challenge_id == challenge["id"]
+            )
+        )
+        return make_challenge_record(challenge, participants)
+
+
+def list_challenges_helper(offset, limit, participant_clause,
+                           where_clause, order_clause):
+    with model.engine.connect() as conn:
+        query = sqlalchemy.sql.select([
+            model.challenges.c.id,
+            model.challenges.c.created,
+            model.challenges.c.finished,
+            model.challenges.c.num_games,
+            model.challenges.c.issuer,
+            model.challenges.c.winner,
+        ]).select_from(model.challenges.join(
+            model.challenge_participants,
+            (model.challenges.c.id == model.challenge_participants.c.challenge_id) &
+            participant_clause
+        )).where(
+            where_clause
+        ).order_by(
+            *order_clause
+        ).offset(offset).limit(limit).reduce_columns()
```

Comment:
Take the following with the caveat that I may be seeing this incorrectly here since I'm manually adding the challenge data to my database to test with.

In the case of no extra participant_clause it seems like this query will return every challenge multiple times, i.e. once for each participant because of the join. But rather than fix this query to only return one row per challenge, it might be better to expand it to include the participant columns and then format the json result from there. This would do away with the next section that is querying the database for every participant for each challenge. That way it would be 1 database query instead of 1+total_number_of_challenge_participants queries. I think this also requires moving the participant_clause from being a join condition to a where condition.

Reply:
You're right about the duplicates. Pushed a fix that should only use one query.